plugins/chain_plugin/get_info_db.cpp

```diff
@@ -0,0 +1,109 @@
+#include <eosio/chain_plugin/get_info_db.hpp>
+#include <eosio/chain/resource_limits.hpp>
+#include <eosio/chain/application.hpp>
+
+#include <shared_mutex>
+
+using namespace eosio;
+using namespace eosio::chain;
+using namespace appbase;
+
+namespace eosio::chain_apis {
+   /**
+    * Implementation details of the get_info results
+    */
+   struct get_info_db_impl {
+      get_info_db_impl(const chain::controller& controller, bool get_info_enabled)
+         : controller(controller)
+         , get_info_enabled(get_info_enabled)
+         , chain_id(controller.get_chain_id())
+         , version_hex(fc::itoh(static_cast<uint32_t>(app().version())))
+         , version_string(app().version_string())
+         , full_version_string(app().full_version_string()) {}
+
+      // A handle to the controller.
+      const chain::controller& controller;
+
+      // Indication whether get_info RPC is enabled.
+      const bool get_info_enabled = false;
+
+      // Cache to store the current get_info results in respect to latest accepted block.
+      mutable std::shared_mutex      rw_mutex;
+      get_info_db::get_info_results  cached_results;
+
+      // Values are never changed
+      const chain_id_type chain_id;
+      const std::string   version_hex;
+      const std::string   version_string;
+      const std::string   full_version_string;
+
+      // Called on accepted_block signal.
+      void on_accepted_block() {
+         try {
+            if (!get_info_enabled) {
+               return;
+            }
+
+            std::unique_lock write_lock(rw_mutex);
```

Comment:
This can be moved down right before the `cached_results = `

Reply:
Thanks. New implementation does not use mutex any more.